### ISSUE TITLE
numpy.recarray.sort segfaults on Windows

### ISSUE THREAD

**Issue by cgohlke** [numpy/numpy#6922]
Using msvc builds of numpy 1.10.2 on Windows, the following crashes on all Python versions:
```
import numpy
a = numpy.recarray(3, dtype=[('object', 'O'), ('int', '<i2')])
a.sort()
```
Call Stack:
```
>	multiarray.pyd!OBJECT_copyswap(_object * * dst=0x0000000003165150, _object * * src=0x000000000325449c, int __NPY_UNUSED_TAGGEDswap=0, void * __NPY_UNUSED_TAGGEDarr=0x00000000005ef1b0)  Line 2037 + 0x21 bytes	C
 	multiarray.pyd!VOID_compare(char * ip1=0x00000000030acb20, char * ip2=0x000000000325449c, tagPyArrayObject * ap=0x0000000003667208)  Line 2812	C
 	multiarray.pyd!npy_quicksort(void * start=0x0000000003254490, __int64 num=3, void * varr=0x0000000003667208)  Line 427 + 0x28 bytes	C
 	multiarray.pyd!_new_sortlike(tagPyArrayObject * op=0x0000000003667208, int axis=0, int (void *, __int64, void *)* sort=0x00007ffceb084940, int (void *, __int64, __int64, __int64 *, __int64 *, void *)* part=0x0000000000000000, __int64 * kth=0x0000000000000000, __int64 nkth=0)  Line 871 + 0x1c bytes	C
 	multiarray.pyd!PyArray_Sort(tagPyArrayObject * op=0x0000000003667208, int axis=0, NPY_SORTKIND which=NPY_QUICKSORT)  Line 1147	C
 	multiarray.pyd!array_sort(tagPyArrayObject * self=0x0000000003667208, _object * args=0x0000000002430048, _object * kwds=0x0000000000000000)  Line 1159 + 0x16 bytes	C
 	python34.dll!PyCFunction_Call(_object * func=0x0000000000000000, _object * arg=0x00000000024bee30, _object * kw=0x0000000003608388)  Line 101	C
 	python34.dll!call_function(_object * * * pp_stack=0x00000000005efad9, int oparg=131)  Line 4241	C
 	python34.dll!PyEval_EvalFrameEx(_frame * f=0x0000000002932438, int throwflag=0)  Line 2840	C
 	python34.dll!PyEval_EvalCodeEx(_object * _co=0x0000000000000000, _object * globals=0x0000000000000000, _object * locals=0x0000000002a5e810, _object * * args=0x00000000024d00d0, int argcount=0, _object * * kws=0x0000000000000000, int kwcount=0, _object * * defs=0x0000000000000000, int defcount=0, _object * kwdefs=0x0000000000000000, _object * closure=0x0000000000000000)  Line 3588 + 0xa bytes	C
 	python34.dll!PyEval_EvalCode(_object * co=0x0000000002ab5cb0, _object * globals=0x0000000002afe728, _object * locals=0x000000000292e848)  Line 781	C
 	python34.dll!run_mod(_mod * mod=0x0000000002ab5cb0, _object * filename=0x00000000024b71d0, _object * globals=0x00000000024b71d0, _object * locals=0x0000000000000000, PyCompilerFlags * flags=0x00000000005efea0, _arena * arena=0x00000000024b71d0)  Line 2176	C
 	python34.dll!PyRun_FileExFlags(_iobuf * fp=0x0000000002a5bd91, const char * filename_str=0x0000000002a5bd90, int start=-1, _object * globals=0x0000000002930c78, _object * locals=0x000000000292e848, int closeit=1, PyCompilerFlags * flags=0x00000000005efea0)  Line 2128 + 0x20 bytes	C
 	python34.dll!PyRun_SimpleFileExFlags(_iobuf * fp=0x00000000005efea0, const char * filename=0x0000000002a5bd90, int closeit=6225568, PyCompilerFlags * flags=0x00000000005efea0)  Line 1602 + 0x2b bytes	C
 	python34.dll!PyRun_AnyFileExFlags(_iobuf * fp=0x0000000002a5bd70, const char * filename=0x00000000024b2936, int closeit=0, PyCompilerFlags * flags=0x00000000024b2936)  Line 1288	C
 	python34.dll!run_file(_iobuf * fp=0x00000000ffffffff, const wchar_t * filename=0x0000000000000000, PyCompilerFlags * p_cf=0x0000000000000000)  Line 319 + 0x18 bytes	C
 	python34.dll!Py_Main(int argc=0, unsigned short * * argv=0x0000014c00000000)  Line 751 + 0x15 bytes	C
 	python.exe!__tmainCRTStartup()  Line 552 + 0x19 bytes	C
 	kernel32.dll!BaseThreadInitThunk()  + 0x22 bytes	
 	ntdll.dll!RtlUserThreadStart()  + 0x34 bytes	

```
Matplotlib's [date_index_formatter.py](https://github.com/matplotlib/matplotlib/blob/master/examples/api/date_index_formatter.py) example segfaults for the same reason.

This could be an alignment issue. Using  `dtype=[('object', 'O'), ('int', '<i4')]` does not segfault on 32-bit Python, only 64-bit. Using `dtype=[('object', 'O'), ('int', '<i8')]` does not segfault.

**Comment by charris** [comment]
Is this new?

**Comment by charris** [comment]
@ahaldane This might be related to your work.

**Comment by charris** [comment]
@cgohlke Yes, looks like alignment or size. I can reproduce this on linux.

**Comment by charris** [comment]
The bad commit is 23901aa4 . @jaimefrio This is yours.
```
commit 23901aa4eaeb04d52bbf1869d34633510e9ffefb
Author: Jaime Fernandez <jaime.frio@gmail.com>
Date:   Sat May 30 18:16:10 2015 -0700

    MANT: Style clean-up in VOID_compare
    
    Avoids modifying the dtype of the array.
    
    Also removed explicit memcpy calls, let copyswap take care of it.
```

**Comment by charris** [comment]
@jaimefrio Decided that reversion is the quick way forward here. Feel free to resubmit the cleanup if you track down the problem.

**Comment by charris** [comment]
Fixed by #6924.
